resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1182,7 +1189,112 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                                                finish = true;
+                                            } finally {
+                                                ReferenceCountUtil.release(msg);
+                                            }
+                                        }
+
+                                        @Override
+                                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                                            if (!finish) {
+                                                if (logger.isDebugEnabled()) {
+                                                    logger.debug("{} Error during processing response: TCP [{}: {}]",
+                                                            ctx.channel(), queryId,
+                                                            ctx.channel().remoteAddress(), cause);
+                                                }
+                                                // TCP fallback failed, just use the truncated response as
+                                                qCtx.finish(res);
+                                            }
+                                        }
+                                    });
+                                }
+                            });
+                    bs.connect(res.sender()).addListener(new ChannelFutureListener() {
+                        @Override
+                        public void operationComplete(ChannelFuture future) {
+                            if (future.isSuccess()) {
+                                final Channel channel = future.channel();
+
+                                Promise<AddressedEnvelope<DnsResponse, InetSocketAddress>> promise =
+                                        channel.eventLoop().newPromise();
+                                new TcpDnsQueryContext(DnsNameResolver.this, channel,
+                                        (InetSocketAddress) channel.remoteAddress(), qCtx.question(),
+                                        EMPTY_ADDITIONALS, promise).query(true, future.channel().newPromise());
+                                promise.addListener(
+                                        new FutureListener<AddressedEnvelope<DnsResponse, InetSocketAddress>>() {
+                                    @Override
+                                    public void operationComplete(
+                                            Future<AddressedEnvelope<DnsResponse, InetSocketAddress>> future) {
+                                        if (future.isSuccess()) {
+                                            qCtx.finish(future.getNow());
+                                        } else {
+                                            // TCP fallback failed, just use the truncated response.
+                                            qCtx.finish(res);
+                                        }
+
+                                        channel.close();
```

Comment:
nit: Perhaps better closing before calling finish() so that the connection is cleaned up a little bit sonner?

Reply:
sure why not :)